### ISSUE TITLE
Unification of promise and stream interfaces

### ISSUE THREAD

**Issue by bisubus** [sindresorhus/got#211]
Promise and stream interfaces have the benefits  for buffered usage they don't share with each other. Default (promise) mode has `json` option and seamless promise output. Streamung (`stream`) mode has the access to events and `req` object - besides streaming itself.

The last use for `got` that I've encountered supposed to return a promise of parsed JSON result, but also to have access to `req.abort` (`request` event) and last requested URL (`redirect` event). Sounds like a simple task. This results in creating a hacky 'convenience wrapper' that takes special care of `json` option [because it makes streaming mode throw up](https://github.com/sindresorhus/got/blob/v6.3.0/index.js#L134-L136), recreates a considerable part of [private `asPromise`](https://github.com/sindresorhus/got/blob/v6.3.0/index.js#L80).

Can we have best of both worlds and get a cancellable, hookable promise?

**Comment by floatdrop** [comment]
(I was writing long answer, but my browser crashed)

Long story short: it is quite a hard problem to merge Stream API with Promise API. Mostly because Promises are lacking of `cancel` and `progress` features.

There is a discussion about [aborting requests](https://github.com/whatwg/fetch/issues/27) in `fetch` library (that ends up in [cancelable promises](https://github.com/domenic/cancelable-promise) proposal), so we could look in this way.

**Comment by bisubus** [comment]
@floatdrop I feel sorry for your long comment, the short one seems fine too.

Thanks for the reference, good to know. Cancelable promises seem to be a long run, what about [Bluebird cancellation](http://bluebirdjs.com/docs/api/cancellation.html) at this moment? It is quite conventional in Node land and nonintrusive (third optional argument in promise constructor). The promise magically gets `cancel` method.

This does not solve the other part, hooks. The promise still has to be augmented with additional methods (`on`?) in this case.

A good example is `superagent`, it went the opposite direction. It returns an augmented, thenable stream. Which may be inconvenient, too, because it should be converted to a real promise.

**Comment by piranna** [comment]
First we should define why we want the promises. Got seems to resolve them when the transfer has finished and you have downloaded the full content body, while I would find more interesting that it resolve at the same time the `response` event is dispatched and give a stream ready to be read... Problem of the first approach is that the content will be downloaded to ram, making it a problem for big files.

**Comment by AlexTes** [comment]
I don't think we should stretch promises to be more than they are. I'm already skeptical the `cancel` bit was a good idea, but the next release covers your original example. If you care about the events, why not use the event emitter the stream mode returns? You cited the JSON option not working, but adding your own parsing should be a small effort with the many helpful libs out there.

Is this still a standing request, and if so, what concretely would you like to see added beyond cancellation?

**Comment by kevva** [comment]
> Problem of the first approach is that the content will be downloaded to ram, making it a problem for big files.

That's why we have to streaming interface. Promises usually resolves when the operation is done.

**Comment by bisubus** [comment]
@AlexTes Thanks for the up, I'm happy that there's cancelling at last. Yes, I still have use case I've described in the OP. Streams are must have for big things, but I never actually needed streaming in Got to decrease RAM footprint, all data is small enough to be efficiently handled with promises. 

Sure, I can use stream if I need to get info on redirects with `redirect` hook (and I need it). But what's the point to use Got at all if this results in usual EE mess instead of neat promise async/await flow control? At this point `http` can be used directly and parsed with JSON.stringify. Considering that it is supposed to be *A **nicer** interface to the built-in http module*, I don't think that there's something 'nice' in a situation like that.

For my case I ended up with customized 'convenience wrapper' over another 'convenience wrapper'. And I'm feeling that the library could cover more use cases without adding more complexity. Considering that Got is promise-first and it already uses streams under the hood, it would make sense to just expose used event emitter on the promise, as `stream` prop or `toStream` method... case solved.
